src/common/lru.c

```diff
@@ -35,304 +46,257 @@
+					 max(4, bits - 3),
 					 NULL, &lru_ops,
-					 &lru_cache->dlc_htable);
+					 &lcache->dlc_htable);
 	if (rc)
-		D_GOTO(exit, rc = -DER_NOMEM);
+		D_GOTO(out, rc);
 
 	if (bits >= 0)
-		lru_cache->dlc_csize = (1 << bits);
+		lcache->dlc_csize = (1U << bits);
 	else /* disable LRU */
-		lru_cache->dlc_csize = 0;
+		lcache->dlc_csize = 0;
 
-	lru_cache->dlc_ops = ops;
-
-	D_INIT_LIST_HEAD(&lru_cache->dlc_idle_list);
-	D_INIT_LIST_HEAD(&lru_cache->dlc_busy_list);
-
-	*lcache = lru_cache;
-exit:
-	if (rc != 0)
-		D_FREE(lru_cache);
+	lcache->dlc_count = 0;
+	lcache->dlc_ops = ops;
 
+	*lcache_pp = lcache;
+	lcache = NULL;
+out:
+	if (lcache != NULL)
+		D_FREE(lcache);
 	return rc;
 }
 
 void
 daos_lru_cache_destroy(struct daos_lru_cache *lcache)
 {
+	if (lcache == NULL)
+		return;
 
 	D_DEBUG(DB_TRACE, "Destroying LRU cache\n");
-	/**
-	 * Cannot destroy if either lcache is NULL or
-	 * if there are busy references.
-	 */
-	D_DEBUG(DB_TRACE, "refs_held :%u\n", lcache->dlc_busy_nr);
-	D_ASSERTF(lcache->dlc_busy_nr == 0, "busy=%d", lcache->dlc_busy_nr);
-
 	d_hash_table_debug(&lcache->dlc_htable);
 	d_hash_table_destroy_inplace(&lcache->dlc_htable, true);
 	D_FREE(lcache);
 }
 
-void
-daos_lru_cache_evict(struct daos_lru_cache *lcache,
-		     daos_lru_cond_cb_t cond, void *args)
-{
-	struct daos_llink *llink;
-	struct daos_llink *tmp;
-	unsigned int	   cntr;
-
-	cntr = 0;
-	d_list_for_each_entry(llink, &lcache->dlc_busy_list, ll_qlink) {
-		if (cond == NULL || cond(llink, args)) {
-			/* will be evicted later in daos_lru_ref_release */
-			daos_lru_ref_evict(llink);
-			cntr++;
-		}
-	}
-	D_DEBUG(DB_TRACE, "Marked %d busy items as evicted\n", cntr);
-
-	cntr = 0;
-	d_list_for_each_entry_safe(llink, tmp, &lcache->dlc_idle_list,
-				   ll_qlink) {
-		if (cond == NULL || cond(llink, args)) {
-			d_list_del_init(&llink->ll_qlink);
-			d_hash_rec_delete_at(&lcache->dlc_htable,
-					     &llink->ll_hlink);
-			lcache->dlc_idle_nr--;
-			cntr++;
-		}
-	}
-	D_DEBUG(DB_TRACE, "Evicted %d items from idle list\n", cntr);
-}
+struct lru_evict_arg {
+	daos_lru_cond_cb_t	 cb;
+	void			*arg;
+	d_list_t		 list;
+};
 
-static struct daos_llink *
-lru_fast_search(struct daos_lru_cache *lcache, d_list_t *head,
-		void *key, unsigned int key_size)
+static int
+lru_evict_cb(d_list_t *link, void *arg)
 {
-	struct daos_llink *llink;
-
-	if (d_list_empty(head))
-		return NULL;
-
-	llink = d_list_entry(head->next, struct daos_llink, ll_qlink);
-	if (llink->ll_evicted)
-		return NULL;
-
-	if (llink->ll_ops->lop_cmp_keys(key, key_size, llink)) {
-		D_DEBUG(DB_TRACE, "Found item on the %s list.\n",
-			head == &lcache->dlc_busy_list ? "busy" : "idle");
+	struct daos_llink *llink = link2llink(link);
+	struct lru_evict_arg *cb_arg = arg;
 
-		llink->ll_ref++; /* +1 for caller */
-		return llink;
-	}
-	return NULL;
+	if (cb_arg->cb == NULL || cb_arg->cb(llink, cb_arg->arg))
+		d_list_add(&llink->ll_qlink, &cb_arg->list);
```

Comment:
Does this need to be d_list_move ?   The eviction callback is invoked from d_hash_table_traverse.   Is it not still in the hash table at this point?

Reply:
No. This is **ll_qlink**. This is a different link that used only for this traverse callback and adding to list. To store into hash table is used **ll_link**.